lib/shared/sdk/usbboot/index.js

```diff
@@ -0,0 +1,318 @@
+
+    // We get this error message when reading a command
+    // from the device when the communication has ended
+    .catch({
+      message: 'LIBUSB_TRANSFER_STALL'
+    }, (error) => {
+      debug(`Got ${error.message} when reading a command, assuming everything is done`)
+      return {
+        command: protocol.FILE_MESSAGE_COMMAND_DONE
+      }
+    })
+
+    .then((fileMessage) => {
+      debug(`Received message: ${fileMessage.command} -> ${fileMessage.fileName}`)
+
+      if (fileMessage.command === protocol.FILE_MESSAGE_COMMAND_DONE) {
+        debug('Done')
+        return Bluebird.resolve()
+      }
+
+      return Bluebird.try(() => {
+        if (fileMessage.command === protocol.FILE_MESSAGE_COMMAND_GET_FILE_SIZE) {
+          debug(`Getting the size of ${fileMessage.fileName}`)
+          const fileSize = _.get(files, [ fileMessage.fileName, 'length' ])
+
+          if (_.isNil(fileSize)) {
+            debug(`Couldn't find ${fileMessage.fileName}`)
+            debug('Sending error signal')
+            return protocol.sendErrorSignal(device)
+          }
+
+          debug(`Sending size: ${fileSize}`)
+          return protocol.sendBufferSize(device, fileSize)
+        }
+
+        if (fileMessage.command === protocol.FILE_MESSAGE_COMMAND_READ_FILE) {
+          debug(`Reading ${fileMessage.fileName}`)
+          const fileBuffer = _.get(files, [ fileMessage.fileName ])
+
+          if (_.isNil(fileBuffer)) {
+            debug(`Couldn't find ${fileMessage.fileName}`)
+            debug('Sending error signal')
+            return protocol.sendErrorSignal(device)
+          }
+
+          return protocol.write(device, endpoint, fileBuffer)
+        }
+
+        return Bluebird.reject(new Error(`Unrecognized command: ${fileMessage.command}`))
+      }).then(() => {
+        debug('Starting again')
+        return startFileServer(device, endpoint, files)
+      })
+    })
+}
+
+exports.flash = (device, options = {}) => {
+  // We need to open the device in order to access _configDescriptor
+  debug(`Opening device: ${device.name}`)
+  device.open()
+
+  // Handle 2837 where it can start with two interfaces, the first
+  // is mass storage the second is the vendor interface for programming
+  const addresses = {}
+  /* eslint-disable no-underscore-dangle */
+  if (device._configDescriptor.bNumInterfaces === USB_ENDPOINT_INTERFACES_SOC_BCM2835) {
+  /* eslint-enable no-underscore-dangle */
+    addresses.interface = 0
+    addresses.endpoint = 1
+  } else {
+    addresses.interface = 1
+    addresses.endpoint = 3
+  }
+
+  const deviceInterface = device.interface(addresses.interface)
+  debug(`Claiming interface: ${addresses.interface}`)
+  deviceInterface.claim()
+
+  const endpoint = deviceInterface.endpoint(addresses.endpoint)
+
+  const serialNumberIndex = device.deviceDescriptor.iSerialNumber
+  debug(`Serial number index: ${serialNumberIndex}`)
+
+  if (serialNumberIndex === USB_DESCRIPTOR_NULL_INDEX) {
```

Comment:
As far as I can see (I ran a couple of tests), an index of 0 means there is no data to fetch. Maybe `USB_DESCRIPTOR_NO_DATA_INDEX` makes it better?